Update button for Stencil 2.0 and Adaptive 2.1

dennislzf opened the pull request:
Status: Ready for review
Reviewers: @jeffkamo @avelinet
## Changes
- Created a button stencil for adaptive 2.1 and stencil 2.0
## Jira Tickets:
- [ ] [STEN-24](https://mobify.atlassian.net/browse/STEN-24)
## How to Test
- Clone the repo
- Run `npm install` and `bower install`
- Run `grunt serve`
- Visit `http://localhost:3000/tests/visual/`

dzlau committed:
I forgot to commit

dzlau committed:
added stencil button

dzlau committed:
added some readme

dennislzf commented:
Also before we continue. What do you guys think about the bootstrapesque of predefined classes. I added it for ease of use, or should we always make user define the colours?

dzlau committed:
added scss vars to file

dzlau committed:
added button size and fixed stupid bugs

dzlau committed:
removed some test.dust stuff

dzlau committed:
added auto escaping to element

dzlau committed:
fixed naming and updated readme

dzlau committed:
fixed naming and updated readme

dzlau committed:
forgot to add in defaults

dzlau committed:
added tests

avelinet commented:
@dennislzf thanks for starting on this. In response to your comment above about classes, in general for stencil components we want to provide the bare minimum and any customization would be built on top it based on project needs. We want to avoid bloating the components with too many options that may not be needed.

We also want to limit the configurable variables to the bare minimum for the same reasons. So there's lots of opportunity here to pare things down, like removing the button size options and text alignment options.

avelinet commented:
It would be great to see this built from our existing button component https://github.com/mobify/stencil/tree/master/dist/components/button

dennislzf commented:
@avelinet oops definitely should've looked for 1.0 stencil first.. will update to match

dzlau committed:
changed it to be more like stencil 1.0

ry5n commented on _button.scss:
```diff
@@ -0,0 +1,71 @@
+
+$button__font-color: inherit !default;
+$button__font-family: inherit !default;
+$button__font-size: inherit !default;
```
I’m not sure we get a lot of value out of any of these variables. They all target a specific css property, and it’s not a lot of extra CSS to write project-level styles to override these. I think until we have a compelling need, we should keep configurable variables to a minimum.

That said, I think we may need to add one variable to stencil-variables to accomplish the goals for this component (see below)

ry5n commented on _button.scss:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    user-select:none;
+    -webkit-appearance: none;
```
There is nothing special about the design of this component, except for one thing: the ability for the tap target to be bigger than the visible size of the button. The challenge with this is that ideally the tap size would not affect layout, allowing things to be a little more compact without impacting usability.

The styles here get us halfway there: they allow the button to _potentially_ be taller than the visible portion, but the tappable portion controls layout, rather than the visible portion.

I wasn’t sure how to do this myself, either, so [I made a test page](http://codepen.io/ry5n/pen/WvLJrg?editors=110). My first idea was to use generated content, which seems to work just fine (verified in Chrome/Android, iOS Safari 7+). To ensure consistency of the visual height across components, we’ll want to add a variable for this to the stencil-variables repo.

ry5n commented on _button.scss:
```diff
@@ -0,0 +1,71 @@
+    line-height: normal;
+    white-space: normal;
+
+    span {
```
An element selector for the visual part of the button might be OK, since the dust template can mostly enforce the DOM structure. However, if we do, it should definitely be `.c-button > span`, otherwise any other descendants (like `<span class="c-icon">` will end up getting messed up. If we want to be strict, this element should get a class and be styled with a class selector. Just not sure what to call it – probably `c-button__inner` even though names like that make me cringe.

ry5n replied:
Also, this selector should certainly come after the `user-select` and `-webkit-appearance` properties. I would argue that it should really be brought out into its own block as `.c-select > span` to emphasize that it’s a sub-component.

ry5n commented on _button.scss:
```diff
@@ -0,0 +1,71 @@
+        border-radius: $button__border-radius;
+        padding: $button__padding;
+
+        display:inline-block;
```
Once everything is sorted (not right now), you’ll want to check [CST’s preferred property sort order](https://github.com/mobify/mobify-code-style/blob/37a9e8a11b875eba8d43d239aff4b6aeed44f933/css/.scss-lint.yml#L290).

ry5n replied:
I’d like to make the visible part of this flexed so that:
- we can set a min-height to the desired visual height
- label and other content (e.g. Icon component) can be centered vertically and horizontally
- label can wrap to two lines if needed

See [my test implementation here](http://codepen.io/ry5n/pen/WvLJrg?editors=110).

ry5n commented on _button.scss:
```diff
@@ -0,0 +1,71 @@
+    &:disabled,
+    &.c-button__is-disabled{
+        opacity: 0.6;
+        pointer-events:none;
```
Good catch – pointer events should prevent activating disabled links. (thumsup)

ry5n commented on button.dust:
```diff
@@ -0,0 +1,13 @@
+                <span>{?text}{text}{/text}</span>
+            </button>
+        {/href}
+    {/element}
```
I’d recommend a number of changes here:
- The `{element|s}` usage looks like it comes from the Select component. For that component, I think it’s sensible to support the use case of providing the entire `<select>` markup including options in one chunk (`element`). Here, that doesn’t make sense, since we’re creating our own button or anchor tag. All we need is the content between `{:else}` and `{/element}`.
- We should use the main component body instead of the `text` parameter for the button content, since that will be a more natural equivalent to what you’d write in HTML (we want to be able to write `{@c-button}Hello World{/c-button}`).
- I’d like to move towards standardizing the order of common html attributes, with `class` first, followed by `id`.
- The disabled state class should be `c--is-disabled` since it’s a modifier rather than a sub-component. In addition, the class should be applied to both the button and anchor elements so that they can be styled using the same selector.

That would leave us with the following template:

``` html
{?href}
    <a class="c-button {class} {?disabled}c--is-disabled{/disabled}"
        {?id}id="{id}"{/id}
        href="{href}"
    >
        <span>{body}</span>
    </a>
{:else}
    <button class="c-button {class} {?disabled}c--is-disabled{/disabled}"
        {?id}id="{id}"{/id}
        type="{?type}{type}{:else}button{/type}"
        {?disabled}disabled{/disabled}
    >
        <span>{body}</span>
    </button>
{/href}
```

ry5n commented:
This is shaping up, we’re better than halfway there I think. Good work @dennislzf! Let me know if you have questions about any of my comments, I’m happy to discuss them.

Oh, almost forgot: we’ll need to update the schema and the readme too, but can do that once this is ready to ship.

ry5n commented on tests/visual/tests.dust:
```diff
@@ -0,0 +1,67 @@
+
+    {/c-spec__test}
+{/c-spec}
+
```
These tests look quite solid. Good work translating the design notes into this spec.

dzlau committed:
added ryans suggestions

dennislzf commented:
added @ry5n suggestions. Also added a variable to stencil-variables. Not sure if this is a good place to put it. https://github.com/mobify/stencil-variables/pull/1

ry5n committed:
Add some missing repo files.

ry5n committed:
Tweak a few things.

- Temporarily change version of stencil-variables dependency.
- Improve formatting and add comments.
- Fix issue with tap-target positioning.
- Remove some styles such as user-select: none, cursor: pointer and white-space: normal.
- Fix some redundant style declarations (appearance and -webkit-appearance).
- Add default padding and border to visible button element.
- Add focus styles.
- Add test for independent control of tap target size.

ry5n commented:
One thing I’d like to evaluate is the cost/benefit on having the internal span. It’s adding a significant amount of complexity and the only real reason we need it is to avoid bugs with flexing `<button>` elements in older browsers. Flex itself is used so that we can control the min-height of the button independent of padding.

So the span might not be worth it. We can still create a larger tap target with a `::before` pesudo-element and style the component root like you’d expect.

ry5n committed:
Update README.

- Update template params.
- Add template bodies.
- Other tweaks

highruned committed:
Switch back to stencil-variables main so we can install properly in projects

jeffkamo committed:
Fix tests and update markup and some classes

jeffkamo committed:
Fix child selector to only apply to disabled buttons

jeffkamo committed:
Add class onto inner

jeffkamo committed:
Remove component attribute

This was breaking sites, because Adaptive tries to initialize the
component but finds no ui script, thus breaking stuff.

jeffkamo committed:
Remove placeholder classes

Allowing these classes to be extended was probably a mistake

Commit:
Change class to innerClass

Commit:
Merge pull request #2 from mobify/dev-innerClass

Change class to innerClass
